[Safe CPP] Address Warnings in RenderTableCol.cpp

stwrt opened the pull request:
#### 548dd5dbfa84c67a969c9adeb6861633b12f4d39
<pre>
[Safe CPP] Address Warnings in RenderTableCol.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=293051">https://bugs.webkit.org/show_bug.cgi?id=293051</a>
<a href="https://rdar.apple.com/problem/151391328">rdar://problem/151391328</a>

Reviewed by Ryosuke Niwa.

Address safe cpp warnings in RenderTableCol.cpp.

To address the warnings I was required to implement downcast support into
CheckedPtrs. The implementation is based off of RefPtr&apos;s downcast support with
minor changes to account for differences between classes and implement the static_pointer_cast
directly into the function.

 * Source/WTF/wtf/CheckedPtr.h:
 (WTF::uncheckedDowncast):
 (WTF::downcast):
 (WTF::dynamicDowncast):
 * Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
 * Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
 * Source/WebCore/rendering/AutoTableLayout.cpp:
 (WebCore::AutoTableLayout::fullRecalc):
 * Source/WebCore/rendering/FixedTableLayout.cpp:
 (WebCore::FixedTableLayout::calcWidthArray):
 * Source/WebCore/rendering/RenderElement.h:
 (WebCore::RenderElement::protectedStyle const):
 * Source/WebCore/rendering/RenderTable.cpp:
 (WebCore::RenderTable::updateColumnCache const):
 (WebCore::RenderTable::effectiveIndexOfColumn const):
 (WebCore::RenderTable::offsetWidthForColumn const):
 * Source/WebCore/rendering/RenderTable.h:
 * Source/WebCore/rendering/RenderTableCell.cpp:
 (WebCore::RenderTableCell::logicalWidthFromColumns const):
 (WebCore::RenderTableCell::computeCollapsedStartBorder const):
 (WebCore::RenderTableCell::computeCollapsedEndBorder const):
(WebCore::RenderTableCell::computeCollapsedBeforeBorder const):
 (WebCore::RenderTableCell::computeCollapsedAfterBorder const):
 * Source/WebCore/rendering/RenderTableCol.cpp:
 (WebCore::RenderTableCol::styleDidChange):
 (WebCore::RenderTableCol::updateFromElement):
 (WebCore::RenderTableCol::willBeRemovedFromTree):
(WebCore::RenderTableCol::clippedOverflowRect const):
 (WebCore::RenderTableCol::clearPreferredLogicalWidthsDirtyBits):
 (WebCore::RenderTableCol::table const):
 (WebCore::RenderTableCol::enclosingColumnGroup const):
 (WebCore::RenderTableCol::nextColumn const):
 (WebCore::RenderTableCol::borderAdjoiningCellStartBorder const):
 (WebCore::RenderTableCol::borderAdjoiningCellEndBorder const):
 (WebCore::RenderTableCol::borderAdjoiningCellBefore const):
 (WebCore::RenderTableCol::borderAdjoiningCellAfter const):
 * Source/WebCore/rendering/RenderTableCol.h:
 (WebCore::RenderTableCol::enclosingColumnGroupIfAdjacentBefore const):
 (WebCore::RenderTableCol::enclosingColumnGroupIfAdjacentAfter const):
 * Source/WebCore/rendering/RenderTableSection.cpp:
 (WebCore::RenderTableSection::paintCell):

Canonical link: <a href="https://commits.webkit.org/295455@main">https://commits.webkit.org/295455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dcafb7385d9b6ad19b14ada693917d0ce46b5542

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104443 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24147 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14571 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109650 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55114 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106483 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24534 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32694 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79319 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107449 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19086 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94271 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59645 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18879 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12343 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54481 "Failed to compile WebKit") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/97122 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88579 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12401 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112037 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/103060 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31604 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23284 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88355 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31968 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90504 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88021 "Found 1 new API test failure: /WebKitGTK/TestCookieManager:/webkit/WebKitCookieManager/persistent-storage (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22583 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32907 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10686 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/26079 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31533 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36863 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/127325 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Failed to compile JSC; Compiled JSC (cancelled)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31323 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/127325 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Failed to compile JSC; Compiled JSC (cancelled)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34662 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32883 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/rendering/RenderTableCol.cpp:
```diff
@@ -168,21 +172,25 @@ void RenderTableCol::clearPreferredLogicalWidthsDirtyBits()
         table = table->parent();
-    return dynamicDowncast<RenderTable>(table);
+
+    if (!table)
```
Instead of adding this explicit `nullptr` check you can simply pass `table` to dynamicDowncast below instead.

stwrt replied:
Fixed in latest patch by adding support for downcasting to CheckedPtr.

rniwa commented on Source/WTF/wtf/CheckedPtr.h:
```diff
@@ -238,4 +277,4 @@ template<typename T> using PackedCheckedPtr = CheckedPtr<T, PackedPtrTraits<T>>;
 using WTF::CheckedPtr;
 using WTF::PackedCheckedPtr;
-
+using WTF::static_pointer_cast;
```
I don't think we want to expose this since static_pointer_cast is not safe to use.

rniwa commented on Source/WTF/wtf/CheckedPtr.h:
```diff
@@ -184,10 +184,21 @@ class CheckedPtr {
+template<typename T, typename U = RawPtrTraits<T>, typename X, typename Y>
+inline CheckedPtr<T, U> static_pointer_cast(const CheckedPtr<X, Y>& p)
+{
+    return CheckedPtr<T, U>(static_cast<T*>(p.get()));
```
I think we wanna do this static cast directly in uncheckedDowncast, etc... and suppress the unsafe cast warning on that line with `SUPPRESS_MEMORY_UNSAFE_CAST`

stwrt reviewed (commented):
> I think we wanna do this static cast directly in uncheckedDowncast, etc... and suppress the unsafe cast warning on that line with `SUPPRESS_MEMORY_UNSAFE_CAST` 

Wasn't getting any warnings with my implementation. Is this check needed?

stwrt reviewed (commented):
> I don't think we want to expose this since static_pointer_cast is not safe to use. 

Removed.

rniwa commented on Source/WTF/wtf/CheckedPtr.h:
```diff
@@ -212,6 +212,34 @@ inline bool is(const CheckedPtr<ArgType, ArgPtrTraits>& source)
 }
 
+template<typename Target, typename Source, typename PtrTraits>
+inline CheckedPtr<match_constness_t<Source, Target>> uncheckedDowncast(CheckedPtr<Source, PtrTraits> source)
```
So we need two variants one which takes `const CheckedPtr<Source, PtrTraits>&` which just calls `get` and another which takes `CheckedPtr<Source, PtrTraits>&&` and does `WTFMove`.

rniwa commented on Source/WTF/wtf/CheckedPtr.h:
```diff
@@ -212,6 +212,34 @@ inline bool is(const CheckedPtr<ArgType, ArgPtrTraits>& source)
+}
+
+template<typename Target, typename Source, typename PtrTraits>
+inline CheckedPtr<match_constness_t<Source, Target>> downcast(CheckedPtr<Source, PtrTraits> source)
```
Ditto.

rniwa commented on Source/WTF/wtf/CheckedPtr.h:
```diff
@@ -212,6 +212,34 @@ inline bool is(const CheckedPtr<ArgType, ArgPtrTraits>& source)
+}
+
+template<typename Target, typename Source, typename TargetPtrTraits = RawPtrTraits<Target>, typename SourcePtrTraits>
+inline CheckedPtr<match_constness_t<Source, Target>, TargetPtrTraits> dynamicDowncast(CheckedPtr<Source, SourcePtrTraits> source)
```
Ditto.

rniwa commented on Source/WTF/wtf/CheckedPtr.h:
```diff
@@ -212,6 +212,62 @@ inline bool is(const CheckedPtr<ArgType, ArgPtrTraits>& source)
 }
 
+template<typename Target, typename Source, typename PtrTraits>
+inline CheckedPtr<match_constness_t<Source, Target>> uncheckedDowncast(CheckedPtr<Source, PtrTraits>& source)
```
This should be `const CheckedPtr<Source, PtrTraits>&`.

rniwa commented on Source/WTF/wtf/CheckedPtr.h:
```diff
@@ -212,6 +212,62 @@ inline bool is(const CheckedPtr<ArgType, ArgPtrTraits>& source)
+}
+
+template<typename Target, typename Source, typename PtrTraits>
+inline CheckedPtr<match_constness_t<Source, Target>> downcast(CheckedPtr<Source, PtrTraits>& source)
```
This should be `const CheckedPtr<Source, PtrTraits>&`

rniwa commented on Source/WTF/wtf/CheckedPtr.h:
```diff
@@ -212,6 +212,62 @@ inline bool is(const CheckedPtr<ArgType, ArgPtrTraits>& source)
+}
+
+template<typename Target, typename Source, typename TargetPtrTraits = RawPtrTraits<Target>, typename SourcePtrTraits>
+inline CheckedPtr<match_constness_t<Source, Target>, TargetPtrTraits> dynamicDowncast(CheckedPtr<Source, SourcePtrTraits>& source)
```
This should be `const CheckedPtr<Source, SourcePtrTraits >&`

stwrt committed:
[Safe CPP] Address Warnings in RenderTableCol.cpp
https://bugs.webkit.org/show_bug.cgi?id=293051
rdar://problem/151391328

Reviewed by Ryosuke Niwa.

Address safe cpp warnings in RenderTableCol.cpp.

To address the warnings I was required to implement downcast support into
CheckedPtrs. The implementation is based off of RefPtr's downcast support with
minor changes to account for differences between classes and implement the static_pointer_cast
directly into the function.

 * Source/WTF/wtf/CheckedPtr.h:
 (WTF::uncheckedDowncast):
 (WTF::downcast):
 (WTF::dynamicDowncast):
 * Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
 * Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
 * Source/WebCore/rendering/AutoTableLayout.cpp:
 (WebCore::AutoTableLayout::fullRecalc):
 * Source/WebCore/rendering/FixedTableLayout.cpp:
 (WebCore::FixedTableLayout::calcWidthArray):
 * Source/WebCore/rendering/RenderElement.h:
 (WebCore::RenderElement::protectedStyle const):
 * Source/WebCore/rendering/RenderTable.cpp:
 (WebCore::RenderTable::updateColumnCache const):
 (WebCore::RenderTable::effectiveIndexOfColumn const):
 (WebCore::RenderTable::offsetWidthForColumn const):
 * Source/WebCore/rendering/RenderTable.h:
 * Source/WebCore/rendering/RenderTableCell.cpp:
 (WebCore::RenderTableCell::logicalWidthFromColumns const):
 (WebCore::RenderTableCell::computeCollapsedStartBorder const):
 (WebCore::RenderTableCell::computeCollapsedEndBorder const):
(WebCore::RenderTableCell::computeCollapsedBeforeBorder const):
 (WebCore::RenderTableCell::computeCollapsedAfterBorder const):
 * Source/WebCore/rendering/RenderTableCol.cpp:
 (WebCore::RenderTableCol::styleDidChange):
 (WebCore::RenderTableCol::updateFromElement):
 (WebCore::RenderTableCol::willBeRemovedFromTree):
(WebCore::RenderTableCol::clippedOverflowRect const):
 (WebCore::RenderTableCol::clearPreferredLogicalWidthsDirtyBits):
 (WebCore::RenderTableCol::table const):
 (WebCore::RenderTableCol::enclosingColumnGroup const):
 (WebCore::RenderTableCol::nextColumn const):
 (WebCore::RenderTableCol::borderAdjoiningCellStartBorder const):
 (WebCore::RenderTableCol::borderAdjoiningCellEndBorder const):
 (WebCore::RenderTableCol::borderAdjoiningCellBefore const):
 (WebCore::RenderTableCol::borderAdjoiningCellAfter const):
 * Source/WebCore/rendering/RenderTableCol.h:
 (WebCore::RenderTableCol::enclosingColumnGroupIfAdjacentBefore const):
 (WebCore::RenderTableCol::enclosingColumnGroupIfAdjacentAfter const):
 * Source/WebCore/rendering/RenderTableSection.cpp:
 (WebCore::RenderTableSection::paintCell):

Canonical link: https://commits.webkit.org/295455@main

nt1m commented on Source/WebCore/rendering/RenderTableCol.cpp:
```diff
@@ -38,6 +38,8 @@
 #include "RenderTable.h"
 #include "RenderTableCaption.h"
 #include "RenderTableCell.h"
+#include "rendering/RenderObject.h"
+#include "rendering/style/RenderStyle.h"
```
This should just be:

`#include "RenderObject.h"` and `#include "RenderStyle.h"`

nt1m commented on Source/WebCore/rendering/RenderTableCol.cpp:
```diff
@@ -168,21 +172,22 @@ void RenderTableCol::clearPreferredLogicalWidthsDirtyBits()
 
-    for (auto& child : childrenOfType<RenderObject>(*this))
-        child.setPreferredLogicalWidthsDirty(false);
+    for (CheckedRef<RenderObject> child : childrenOfType<RenderObject>(*this))
```
The type isn't necessary here. `CheckedRef child` is sufficient.

nt1m commented on Source/WebCore/rendering/RenderTableCol.cpp:
```diff
@@ -67,21 +69,23 @@ RenderTableCol::~RenderTableCol() = default;
         return;
-    table->invalidateCollapsedBordersAfterStyleChangeIfNeeded(*oldStyle, style());
+
+    CheckedRef<const RenderStyle> newStyle = style();
```
`CheckedRef newStyle` is fine here. We don't typically specify if something is const

nt1m commented on Source/WebCore/rendering/RenderTableCol.cpp:
```diff
@@ -168,21 +172,22 @@ void RenderTableCol::clearPreferredLogicalWidthsDirtyBits()
+    CheckedPtr table = parent();
     if (table && !is<RenderTable>(*table))
         table = table->parent();
+
     return dynamicDowncast<RenderTable>(table);
 }
 
-RenderTableCol* RenderTableCol::enclosingColumnGroup() const
+CheckedPtr<RenderTableCol> RenderTableCol::enclosingColumnGroup() const
```
I think all of these changes are against the style guide. These should have a separate `checked` method.

cc @cdumez

nt1m commented on Source/WebCore/rendering/RenderTableCol.h:
```diff
@@ -49,12 +51,12 @@ class RenderTableCol final : public RenderBox {
 
-    RenderTableCol* enclosingColumnGroup() const;
-    RenderTableCol* enclosingColumnGroupIfAdjacentBefore() const;
-    RenderTableCol* enclosingColumnGroupIfAdjacentAfter() const;
+    CheckedPtr<RenderTableCol> enclosingColumnGroup() const;
+    CheckedPtr<RenderTableCol> enclosingColumnGroupIfAdjacentBefore() const;
+    CheckedPtr<RenderTableCol> enclosingColumnGroupIfAdjacentAfter() const;
 
     // Returns the next column or column-group.
-    RenderTableCol* nextColumn() const;
+    CheckedPtr<RenderTableCol> nextColumn() const;
```
Same here, these are all against the webkit style afaik.

Below as well

nt1m commented on Source/WebCore/rendering/RenderElement.h:
```diff
@@ -68,6 +69,7 @@ class RenderElement : public RenderObject {
     bool hasInitializedStyle() const { return m_hasInitializedStyle; }
 
     const RenderStyle& style() const { return m_style; }
+    CheckedRef<const RenderStyle> protectedStyle() const { return m_style; }
```
This should be named `checkedStyle()` not `protectedStyle()` since it returns a `CheckedRef`

nt1m commented on Source/WebCore/rendering/RenderTableCol.cpp:
```diff
@@ -214,24 +220,28 @@ RenderTableCol* RenderTableCol::nextColumn() const
 const BorderValue& RenderTableCol::borderAdjoiningCellStartBorder() const
 {
-    return style().borderStart(table()->writingMode());
+    const WritingMode tableWritingMode = table()->writingMode();
+    return protectedStyle()->borderStart(tableWritingMode);
```
Here and below: not sure I understand this change? How is this necessary?
